core/src/main/java/org/frankframework/http/authentication/SamlAssertionOauth.java

```diff
@@ -0,0 +1,257 @@
+		conditions.setAttribute("NotBefore", NotBefore);
+		conditions.setAttribute("NotOnOrAfter", NotOnOrAfter);
+
+		Element audienceRestriction = doc.createElementNS(namespaceURI, "saml2:AudienceRestriction");
+
+		Element audienceElement = doc.createElementNS(namespaceURI, "saml2:Audience");
+		audienceElement.setTextContent(session.getSamlAudience());
+
+		audienceRestriction.appendChild(audienceElement);
+		conditions.appendChild(audienceRestriction);
+		assertion.appendChild(conditions);
+
+		// Add AuthnStatement
+		Element authnStatement = doc.createElementNS(namespaceURI, "saml2:AuthnStatement");
+		authnStatement.setAttribute("AuthnInstant", now);
+		authnStatement.setAttribute("SessionIndex", UUID.randomUUID().toString());
+
+		Element authnContext = doc.createElementNS(namespaceURI, "saml2:AuthnContext");
+
+		Element authnContextClassRef = doc.createElementNS(namespaceURI, "saml2:AuthnContextClassRef");
+		authnContextClassRef.setTextContent("urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport");
+
+		authnContext.appendChild(authnContextClassRef);
+		authnStatement.appendChild(authnContext);
+		assertion.appendChild(authnStatement);
+
+		// Add attributeStatement
+		Element attributeStatement = doc.createElementNS(namespaceURI, "saml2:AttributeStatement");
+
+		Element attribute = doc.createElementNS(namespaceURI, "saml2:Attribute");
+		attribute.setAttribute("Name", "api_key");
+
+		Element attributeValue = doc.createElementNS(namespaceURI, "saml2:AttributeValue");
+		attributeValue.setAttribute("xsi:type", "xs:string");
+		attributeValue.setTextContent(session.getClientId());
+
+		attribute.appendChild(attributeValue);
+		attributeStatement.appendChild(attribute);
+		assertion.appendChild(attributeStatement);
+
+		return doc;
+	}
+
+	private Document signAssertion(Document doc, PrivateKey privateKey, X509Certificate certificate) throws Exception {
+		// Register the "ID" attribute as type ID in the XML document
+		Element rootElement = doc.getDocumentElement();
+		rootElement.setIdAttribute("ID", true);
+
+		// Create the XMLSignature object
+		XMLSignature signature = new XMLSignature(doc, "", XMLSignature.ALGO_ID_SIGNATURE_RSA_SHA256);
+
+		// Add transforms (Canonicalization)
+		Transforms transforms = new Transforms(doc);
+		transforms.addTransform(Transforms.TRANSFORM_ENVELOPED_SIGNATURE); // For signature element
+		transforms.addTransform(Transforms.TRANSFORM_C14N_EXCL_OMIT_COMMENTS); // Canonicalization
+		signature.addDocument("#" + rootElement.getAttribute("ID"), transforms, MessageDigestAlgorithm.ALGO_ID_DIGEST_SHA256);
+
+		// Add the KeyInfo to the Signature
+		signature.addKeyInfo(certificate);
+		signature.addKeyInfo(certificate.getPublicKey());
+
+		// Append the signature to the root element
+		rootElement.appendChild(signature.getElement());
+
+		// Sign the document
+		signature.sign(privateKey);
+
+		return doc;
+	}
+
+	public String documentToString(Document document) throws Exception {
```

Comment:
Heel misschien zou je XmlUtils.nodeToString(.., ..) kunnen gebruiken. Ik zie alleen dat `OMIT_XML_DECLARATION` daar standaard op true staat. Mogelijk maakt dat niet uit.

Reply:
Die twee properties zijn heel erg belangerijk, omdat anders de signature niet meer klopt. Ik had het hier gisteren ook met Tim over en we kwamen tot de conclusie dat er niet echt iets bruikbaars in XmlUtils zat voor mij.